### ISSUE TITLE
Request: Independent window rescale

### ISSUE THREAD

**Issue by abrolag** [fltk/fltk#44]
Please consider enabling changing the size of individual windows, keeping aspect ratio and relative widget locations. We would like to use this for sight impaired people, so that without oversizing the whole application they can zoom in for individual details, then back out again.

I would suggest [ALT] + and [ALT} - for controls.

If this is not considered important enough, are there hooks in the code where we could attempt to arrange this ourselves.

This is for the Yoshimi soft-synth. We try to do as much as we can for disabled musicians.

**Comment by erco77** [comment]
In fltk 1.4.x, [Ctrl]- and [Ctrl]+ does this. Should work with all of the fltk test programs.

**Comment by abrolag** [comment]
Oh that's great news!
Just to be sure I've got this right, so an application with two windows can have one enlarged, while the other one stays the same. Is that right?

**Comment by erco77** [comment]
Oh, *per-window* in a multiwindow app? Offhand I don't think it works that way, no, so your request should stand unless there's something I'm missing.
I believe the Ctrl+/- scaling is for all windows in the app, not just the one you're currently in, the goal being to help the application look right on high dpi displays.

**Comment by MatthiasWM** [comment]
Also, in Apple MacOS, you can scale the entire screen simply by holding Ctrl and dragging the scroll wheel (or the two-finger scroll on the touch pad). It's really a fantastic help for sight impaired people and aging programmers.

**Comment by ManoloFLTK** [comment]
What is currently possible with FLTK 1.4 is to define a global shortcut key
that would enlarge all windows on a screen, and, for example, reset all windows
to their previous size if repeated. 
That is nearly what you asked for except that other windows of the FLTK app
get enlarged too, which may be acceptable since switching is very simple.
The following example defines ctrl-P  as that global shortcut
and have it enlarge all window content 3-fold. That won't work if the ctrl-P
shortcut is used for something else by the FLTK app (say, a menu item shortcut).


1) add this function to your code:
#include <src/Fl_Screen_Driver.H>
int my_scale_handler(int event) {
  if ( event != FL_SHORTCUT || (!Fl::event_ctrl())  ) return 0;
  int key = Fl::event_key() & ~(FL_SHIFT+FL_CTRL);
  if (key != 'p')  return 0;
  static float current_scale = 1;
  static int current_screen = -1;
  if (current_screen == -1 )  {
    current_screen = Fl::first_window()->screen_num();
    current_scale = Fl::screen_scale(current_screen);
    Fl::screen_driver()->rescale_all_windows_from_screen(current_screen, 3);
  }
  else  {
    Fl::screen_driver()->rescale_all_windows_from_screen(current_screen, current_scale);
    current_screen = -1;
  }
  return 1;
}

2) put this statement very early in your program, possibly as 1st statement in main():
  Fl::add_handler(my_scale_handler);

**Comment by ManoloFLTK** [comment]
I believe there's no convenient hook to the code to rescale an FLTK window
individually. That is because FLTK defines a per-display scaling factor
and refers to that value in many places under the logic that all windows
of the same display share the same value of the scaling factor.
That property allows all FLTK apps to become rescalable with FLTK 1.4 without 
change to their source code.

**Comment by abrolag** [comment]
Thanks for the updates. A bit disappointing, but still an impovement on the overall fixed size of fltk V 1.3

**Comment by Albrecht-S** [comment]
I believe that this issue has been discussed and will not be fixed, at least not in the near or middle future.

I suggest to close this issue. It will be closed no later than 1 week from now if there's no further action.

**Comment by erco77** [comment]
We could plan for such changes, perhaps as a discussion on fltk.coredev.
Or an RFE/STR that remains open so the request and implementation suggestions don't get lost,
as I imagine there's several needs, each with its own details. (multi-monitor/screen specific vs. window specific, etc)
I imagine implementation could be tricky, since deparented windows like menus and
dialogs would probably have to be affected too.

**Comment by abrolag** [comment]
This seems like a good compromise. While I appreciate this isn't a high priority for most people, I wouldn't want it to get lost completely.

**Comment by Albrecht-S** [comment]
Okay, points taken. I removed the 'wontfix' label and added 'enhancement' so we can leave this issue open for now and discuss the possible improvements further.

**Comment by abrolag** [comment]
Just as a matter of interest, we've been able to get what we wanted with fltk1.3.5. We set the windows as resizable, and work with a fixed W/H ratio (so circles remain circles), then store the default width. When resizing we multiply text font size by the ratio of the current width against the default.
Menu bars and tabs (with text) don't work correctly so we synthesize these.
If anyone is interested the project is Yoshimi soft-synth and is on here and sourceforge.

**Comment by rageworx** [comment]
> In fltk 1.4.x, [Ctrl]- and [Ctrl]+ does this. Should work with all of the fltk test programs.

Dear @Albrecht-S , 
I read this now, and really super thanks for announcing this feature to test high DPI affects on 96 DPI screen !

**Comment by MatthiasWM** [comment]
The discussion drifted from individual window scaling to a more general approach to scaling UIs.

I would ask the participants to either close this issue, or change the title of this issue, and/or open another issue with the result of this discussion and a concrete feature to implement.

This issue will expire on 23 Nov 2025

**Comment by Albrecht-S** [comment]
I agree with @MatthiasWM that this particular issue should be closed. If anybody wanted to pursue the ideas proposed here, they should open a thread in GitHub Discussions, e.g. in https://github.com/fltk/fltk/discussions/categories/ideas (note that we didn't have discussions yet when this issue was opened).

_**Note to FLTK devs**: please DO NOT convert this issue to a discussion because this would have unwanted side effects (lots of mails...)._

**Comment by MatthiasWM** [comment]
Closed
